### PR TITLE
if aewc is in threat zone move it further away

### LINKED ISSUE #842: ensure AEW&C orbit point is not in threat zone
Follow up for https://github.com/Khopa/dcs_liberation/pull/837.

### PR DESCRIPTION
**Pull request by siKruger**
borrowed from BARCAP generation:

if the AEW&C is in threat zone the waypoint will be moved out of the threat zone

>https://github.com/Khopa/dcs_liberation/issues/842

### CODE REVIEW

File: gen/flights/flightplan.py [hunk 1]
```diff
@@ -1182,14 +1182,24 @@ def racetrack_for_objective(
...
         heading = location.position.heading_between_point(closest_airfield.position)
-        return location.position.point_from_heading(heading, 5000)
+
+        position = ShapelyPoint(
+            self.package.target.position.x, self.package.target.position.y
+        )
+
+        if meters(position.distance(self.threat_zones.all)) == meters(0):
+            distance_to_no_fly = (
+                    meters(position.distance(self.threat_zones.all))
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: This is guaranteed to be zero.

File: gen/flights/flightplan.py [hunk 2]
```diff
@@ -1182,14 +1182,24 @@ def racetrack_for_objective(
...
         return start, end
 
-    @staticmethod
-    def aewc_orbit(location: MissionTarget) -> Point:
+    def aewc_orbit(self, location: MissionTarget) -> Point:
         closest_airfield = location
-        # TODO: This is a heading to itself.
-        # Place this either over the target or as close as possible outside the
-        # threat zone: https://github.com/Khopa/dcs_liberation/issues/842.
         heading = location.position.heading_between_point(closest_airfield.position)
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: This is still guaranteed to be due north, which isn't necessarily away from the threat.

You might need to play with shapely a bit here, but what you need to do is find a direction that is at least likely to be safer. The BARCAP planning isn't perfect either and just backs away from the closest enemy airfield. That might accidentally back into another enemy airfield if the mission location is surrounded, of course, but it's better than nothing. You could copy that here.

File: gen/flights/flightplan.py [hunk 3]
```diff
@@ -1182,14 +1182,24 @@ def racetrack_for_objective(
...
-        return location.position.point_from_heading(heading, 5000)
+
+        position = ShapelyPoint(
+            self.package.target.position.x, self.package.target.position.y
+        )
+
+        if meters(position.distance(self.threat_zones.all)) == meters(0):
+            distance_to_no_fly = (
+                    meters(position.distance(self.threat_zones.all))
+                    - self.doctrine.cap_engagement_range
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: This distance isn't meaningful here. The reason it was used for the BARCAP planning is that its goal is to keep the edge of the engagement zone out of the enemy threat zone. Here we just need to keep the orbit itself safe. Instead, just pick a larger constant than 5nm. 40nm seems reasonable.

**Review by DanAlbert — CHANGES_REQUESTED** [review]

**Comment by siKruger** [comment]
what do you think of this idea?

basically Iam starting from heading 0 and distance 0 and work my way up to find a point that is outside the threat zone.

In my testing I always got a point outside of the threat zones. sometimes they are not optimal but at least they are outside of any zone

**Comment by DanAlbert** [comment]
In general we're trying to move away from brute force point searches, as they're slow and might not find an answer at all (we used to have quite of bit of cases that did things like this).

With shapely you should be able to find the closest point on the edge of the threat zone to the target area. Find that point, find the heading from the airfield to that point, and extend on that bearing? (Unless of course the point near the airfield isn't under threat anyway, in which case just use that point).

**Comment by siKruger** [comment]
> 
> 
> In general we're trying to move away from brute force point searches, as they're slow and might not find an answer at all (we used to have quite of bit of cases that did things like this).
> 
> With shapely you should be able to find the closest point on the edge of the threat zone to the target area. Find that point, find the heading from the airfield to that point, and extend on that bearing? (Unless of course the point near the airfield isn't under threat anyway, in which case just use that point).

Will take a look at this!

**Comment by siKruger** [comment]
I am now searching for the closest boundary point. 

theoretically they are still inside the threat zone, but if we would move them further away they could end up in another threat zone and i guess then we would be back at out old friend mr bruteforce.

should we risk it and move the point further away or just leave it as it is right now?

File: gen/flights/flightplan.py [hunk 4]
```diff
@@ -1182,14 +1182,27 @@ def racetrack_for_objective(
...
+
+            # Heading + Distance to border point
+            heading = location.position.heading_between_point(closest_closest_boundary)
+            distance = location.position.distance_to_point(closest_closest_boundary)
+
+            return location.position.point_from_heading(heading, distance)
+
+        # this Part is fine. No threat zone, just use our point
+        else:
+            return location.position.point_from_heading(heading, 0)
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: `return location` and remove L1189?

File: gen/flights/flightplan.py [hunk 5]
```diff
@@ -1182,14 +1182,27 @@ def racetrack_for_objective(
...
-        return location.position.point_from_heading(heading, 5000)
+    def aewc_orbit(self, location: MissionTarget) -> Point:
+        heading = location.position.heading_between_point(location.position)
+
+        position = ShapelyPoint(
+            self.package.target.position.x, self.package.target.position.y
+        )
+
+        # in threat zone
+        if meters(position.distance(self.threat_zones.all)) == meters(0):
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: `if self.threat_zones.threatened(position)`

If you do that you also don't need to explicitly make the `ShapelyPoint`, it'll accept `location.position` so you can delete L1188-L1190.

File: gen/flights/flightplan.py [hunk 6]
```diff
@@ -1182,14 +1182,27 @@ def racetrack_for_objective(
...
+            # Borderpoint
+            closest_closest_boundary = self.threat_zones.closest_boundary(position)
```
Comment (DanAlbert) [inline gen/flights/flightplan.py]: `closest_boundary`, not `closest_closest_boundary` :)

**Review by DanAlbert — APPROVED** [review]
Suggested a few cleanups. You should be able to merge this yourself now, and aside from the few nits this looks good so feel free to do so :)

I think we may want to buffer the boundary point a bit, but happy to have this merge as-is and so we can start getting some testing on it.

**Pull request merged**